ipv8/peerdiscovery/community.py

```diff
@@ -1,61 +1,106 @@
+from __future__ import annotations
+
 from binascii import unhexlify
 from random import sample
 from time import time
+from typing import TYPE_CHECKING, Iterable, Union, cast
 
-from .churn import DiscoveryStrategy, RandomChurn
-from .payload import (DiscoveryIntroductionRequestPayload, PingPayload, PongPayload, SimilarityRequestPayload,
-                      SimilarityResponsePayload)
-from ..community import Community, DEFAULT_MAX_PEERS
+from ..community import DEFAULT_MAX_PEERS, Community
 from ..keyvault.crypto import default_eccrypto
 from ..lazy_community import PacketDecodingError, lazy_wrapper, lazy_wrapper_unsigned, retrieve_cache
-from ..messaging.payload import IntroductionRequestPayload
+from ..messaging.payload import IntroductionRequestPayload, IntroductionResponsePayload
 from ..messaging.payload_headers import BinMemberAuthenticationPayload, GlobalTimeDistributionPayload
-from ..messaging.serialization import PackError
+from ..messaging.serialization import PackError, Serializable
 from ..peer import Peer
 from ..requestcache import NumberCache, RequestCache
+from .churn import DiscoveryStrategy, RandomChurn
+from .payload import (
+    DiscoveryIntroductionRequestPayload,
+    PingPayload,
+    PongPayload,
+    SimilarityRequestPayload,
+    SimilarityResponsePayload,
+)
 
+if TYPE_CHECKING:
+    from ..types import Address, Endpoint
+    from .network import Network
 
-class PeriodicSimilarity(DiscoveryStrategy):
 
-    def __init__(self, overlay):
+class PeriodicSimilarity(DiscoveryStrategy):
+    """
+    Periodically send a request for similar Communities to a random peer.
+    """
+
+    def __init__(self, overlay: DiscoveryCommunity) -> None:
+        """
+        Create a new strategy to regularly send similarity requests.
+        """
         super().__init__(overlay)
-        self.last_step = 0
+        self.last_step: float = 0
 
-    def take_step(self):
+    def take_step(self) -> None:
+        """
+        Select a random peer, at most every second, to send a similarity request to.
+        """
         now = time()
         if (now - self.last_step < 1.0) or not self.overlay.network.verified_peers:
             return
         self.last_step = now
         with self.walk_lock:
+            self.overlay = cast(DiscoveryCommunity, self.overlay)
```

Comment:
To prevent repetition, it can be moved to the constructor

Reply:
Assuming you mean this:

```patch
diff --git a/ipv8/peerdiscovery/community.py b/ipv8/peerdiscovery/community.py

index 398c28c..0720e85 100644

--- a/ipv8/peerdiscovery/community.py

+++ b/ipv8/peerdiscovery/community.py

@@ -38,6 +38,7 @@ class PeriodicSimilarity(DiscoveryStrategy):

         """

         super().__init__(overlay)

         self.last_step: float = 0

+        self.overlay = cast(DiscoveryCommunity, self.overlay)

 

     def take_step(self) -> None:

         """

@@ -48,7 +49,6 @@ class PeriodicSimilarity(DiscoveryStrategy):

             return

         self.last_step = now

         with self.walk_lock:

-            self.overlay = cast(DiscoveryCommunity, self.overlay)

             self.overlay.send_similarity_request(sample(list(self.overlay.network.verified_peers), 1)[0].address)

```

Gives me an error:

```
ipv8/peerdiscovery/community.py:52: error: "Overlay" has no attribute "send_similarity_request"  [attr-defined]
```